Source/OIDIDToken.m

```diff
@@ -0,0 +1,152 @@
+ */
+
+#import "OIDIDToken.h"
+
+/*! Field keys associated with an ID Token. */
+static NSString *const kIssKey = @"iss";
+static NSString *const kSubKey = @"sub";
+static NSString *const kAudKey = @"aud";
+static NSString *const kExpKey = @"exp";
+static NSString *const kIatKey = @"iat";
+static NSString *const kNonceKey = @"nonce";
+
+#import "OIDFieldMapping.h"
+
+@implementation OIDIDToken
+
+@synthesize header = _header;
+@synthesize claims = _claims;
+@synthesize issuer = _issuer;
+@synthesize subject = _subject;
+@synthesize audience = _audience;
+@synthesize expiresAt = _expiresAt;
+@synthesize issuedAt = _issuedAt;
+@synthesize nonce = _nonce;
+
+- (instancetype)initWithIDTokenString:(NSString *)idToken {
+  self = [super init];
+  NSArray *sections = [idToken componentsSeparatedByString:@"."];
+  if (sections.count > 1) {
+    _header = [[self class] parseJWTSection:sections[0]];
+    _claims = [[self class] parseJWTSection:sections[1]];
+    if (!_header || !_claims) {
+      return nil;
+    }
+
+    [OIDFieldMapping remainingParametersWithMap:[[self class] fieldMap]
+                                     parameters:_claims
+                                       instance:self];
+
+    // Required fields.
+    if (!_issuer || !_audience || !_subject || !_expiresAt || !_issuedAt) {
+      return nil;
+    }
+
+    return self;
+  }
+  return nil;
+}
+
+/*! @brief Returns a mapping of incoming parameters to instance variables.
+    @return A mapping of incoming parameters to instance variables.
+ */
++ (NSDictionary<NSString *, OIDFieldMapping *> *)fieldMap {
+  static NSMutableDictionary<NSString *, OIDFieldMapping *> *fieldMap;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    fieldMap = [NSMutableDictionary dictionary];
+
+    fieldMap[kIssKey] =
+        [[OIDFieldMapping alloc] initWithName:@"_issuer"
+                                         type:[NSURL class]
+                                   conversion:[OIDFieldMapping URLConversion]];
+    fieldMap[kSubKey] =
+        [[OIDFieldMapping alloc] initWithName:@"_subject" type:[NSString class]];
+    fieldMap[kAudKey] =
+        [[OIDFieldMapping alloc] initWithName:@"_audience"
+                                         type:[NSArray class]
+                                   conversion:^id _Nullable(NSObject *_Nullable value) {
+          if ([value isKindOfClass:[NSArray class]]) {
+            return value;
+          }
+          return @[value];
+        }];
+    fieldMap[kExpKey] =
+        [[OIDFieldMapping alloc] initWithName:@"_expiresAt"
+                                         type:[NSDate class]
+                                   conversion:^id _Nullable(NSObject *_Nullable value) {
+          if (![value isKindOfClass:[NSNumber class]]) {
+            return value;
+          }
+          NSNumber *valueAsNumber = (NSNumber *)value;
+          return [NSDate dateWithTimeIntervalSince1970:[valueAsNumber longLongValue]];
```

Comment:
dot notation for properties? (longLongValue below as well)

Reply:
Done.